Fully encapsulate load and save of banlist to disk in DoS Manager

Justaphf opened the pull request:
This builds on top of #619 to fully encapsulate banlist functionality related to saving to and loading from disk.  This allows removal of dependency on CBanDB in net.cpp.

This PR:
1. Moves `DumpBanlist()` from net.cpp to the DoS manager class, including the checks to see if the in-memory banlist is marked as dirty inside `DumpData()`.  Replaces this functionality with a single call to the DoS Manager.
2. Creates a `LoadBanlist()` method in the DoS manager class and moves this functionality from `StartNode` in net.cpp.  Replaces this functionality with a single call to the DoS Manager.
3. Remove bandb.h dependency/include from net.cpp
4. Corrects an issue where the dirty flag was cleared even if the banlist failed to write to disk.
5. Adds additional unit tests related to `DumpBanlist()` and `LoadBanlist()` functionality.

Potential clean-ups:
dff576b is actually a fix to a potential bug I found while moving this functionality from net.cpp and may warrant a separate PR rather than just a separate commit.

Future Tasks:
1. `SetBanned()`, `BannedSetIsDirty()`, and `SetBannedSetDirty()` are now no longer used externally from the DoS manager and can potentially be removed.  This is beneficial as the current implementation (as pulled from net.cpp) takes and releases the `cs_setBanned` lock several times in quick succession by making multiple calls to these methods from within `DumpBanlist()` and `LoadBanlist()`.

Justaphf committed:
Move DumpBanlist from net.cpp to DoS manager

Straight move, further clean-up in subsequent commit.

Justaphf committed:
Encapsulate setBannedIsDirty flag check in

DumpBanlist and remove from DumpData in net.cpp.  This is working
towards the removal of public access to the setBannedIsDirty flag
outside of the manager.

Justaphf committed:
Add LoadBanlist() to dos manager

and remove direct set functionality from StartNode() in net.cpp.
This eliminates the dependency on CBanDB from net.cpp

Justaphf committed:
Add additional DoS Manager unit tests

kyuupichan commented:
I think this makes sense, and breaking out the logic is nice as are the tests.  I'm not sure it will end up being the final location for the ban functionality; perhaps the connection manager will be that place.

I hope @gandrewstone will give guidance on his thoughts on this.

Justaphf commented:
rebase

kyuupichan reviewed (approved):
Again this is just moving code and new tests.  The only change here is a bug fix - previously we would mark the Banned Set clean even if writing it out failed; with these changes we no longer do that.
Looks good to me.

Justaphf commented:
rebase on top of squashed #616

Justaphf commented:
@kyuupichan Looks like since I rebased and squashed after you approved these, they have to be re-approved.  Sorry, missed squashing the node state PR before your review and ended up having to rebase all of them again on top of that squash.

ptschip reviewed (approved):
exellent...approved.

I agree with @kyuupichan  that these changes may migrate , at least some, to the connection manager.  Really in the end we want to get away from cs_main needed for tracking state and also the tracking of blocks in flight should probably end up in the request manager (which will be a rather large task i believe since blocksinflight, thinblocksinflight, IBD, blockdownload timeouts are all related)

Justaphf commented:
rebased on top of comment fix in #619

kyuupichan reviewed (approved):
Excellent; let's get these 5 PRs in.

Justaphf commented:
Rebased on current dev

Justaphf commented:
Let me know if you want me to squash before merge.

ptschip reviewed (approved):
re-approving

gandrewstone commented on src/dosman.cpp:
```diff
@@ -280,3 +280,54 @@ void CDoSManager::Misbehaving(NodeId pnode, int howmuch)
+void CDoSManager::DumpBanlist()
+{
+    // If setBanned is not dirty, don't waste time on disk i/o
+    if (!dosMan.BannedSetIsDirty())
```
don't call the singleton from inside the class.

gandrewstone commented:
also in both IsBanned() functions you don't really need the {} isolating the LOCK() from the rest of the routine since the rest is trivial.

Its inconsistent for the SetBannedSetDirty function and the setBannedIsDirty to be used internally.  If you replaced the variable by the function, the semantics of DumpBanlist() and LoadBanlist() would change because the  cs_setBanned lock would not be taken while setting the variable.  

Should the lock be taken across most of DumpBanlist() and LoadBanlist() anyway?

In reading the code, it looks like in both of these functions there are subtle bugs if the code only locks and unlocks inside each subfunction.  You want the lock held across multiple operations so the operations are atomic relative to other accesses:

In DumpBanList(), if another thread bans a node just after bandb.Write(banmap) but before  SetBannedSetDirty(false), then the banmap will be dirty but the flag set to false.

In LoadBanlist(), another thread could query the banlist between the bandb.Read(banmap) operation and the SweepBanned() operation.  In this situation, I think that we'd return "true" for the node being banned even though the ban is actually expired.

Neither of these threading issues are going to cause a core dump but they are worth fixing since they also clean up the inconsistent use of the function and the variable.  So I'd recommend replacing the SetBannedSetDirty function calls with setBannedIsDirty=false; and locking cs_setBanned at that level.

Justaphf commented:
@gandrewstone Thanks for the comprehensive review!  I have marked this as WIP to correct the internal singleton I missed and to extract the bug fix I included.  It really should have been part of a separate PR so as not to mix move of functionally, bugs and all, with other bug fixes and improvements.  I have intended to add a follow-up PR which would deal with cleaning up the lock usage issues internally, but haven't gotten to it yet.

Final result will be that this PR will move the dump and load functionality line-for-line from net.cpp.  The follow-on PR will then correct the locking issues you mentioned.  Hopefully I'll be able to get something for review in tonight when I get home.

Justaphf commented:
- Rebased on latest dev
- Removed the bugfix so this is purely a move of functionality from net.cpp
- Left addition of unit tests to verify behavior before and after lock clean-up

Leaving as WIP until I have the follow-up PR dealing with the `cs_setBanned` lock clean-ups.

Justaphf commented:
Taking out of WIP now that #662 is in for review.  That PR builds upon this one and addresses the comments from @gandrewstone.  If reviewers determine it to be preferable to have both of these PRs combined into one, I can squash the commits from this PR into 662 for merge.

kyuupichan reviewed (approved):
Approved with fixes in #662
